### DISCUSSION TITLE
Documenting coercion with descriptors

### DISCUSSION THREAD

**Discussion by tuchandra** [pydantic/pydantic#4262] [Pydantic V2]
# Documenting "lax" coercion with descriptors
**tl;dr** The [dataclass_transform mechanism](https://peps.python.org/pep-0681/) that static type checkers use to understand libraries like pydantic supports descriptor-typed fields. This seems like a way for pydantic (v2) to document what types it will attempt to coerce to a target type. I've provided runnable examples below and would love feedback on this.

(All the code in this post is using Python 3.9.9, pydantic v1.9.1, and the VS Code language server Pylance v2022.7.41, which itself uses pyright v1.1.262. I didn't test any of this using mypy, since afaik it doesn't yet support `dataclass_transform`. Finally, I have been a heavy user of pydantic and static typing, but my "type theory" understanding is pretty rough, so there's a good chance I'm misunderstanding something here.)

If you already understand why `SomePydanticModel(field_annotated_as_int="1")` doesn't type check, then you can skip this section ↓

<details>
<summary>Example of pydantic's coercion and how type checkers don't like it</summary>

Because the second half of this post is going to get a little more thorny, I want to make sure that I start from common ground. Here's an example whose behavior is uncontroversial, if undesired.

```py
from __future__ import annotations
from decimal import Decimal
from pydantic import BaseModel

class Model(BaseModel):
    x: float

# this type checks, thanks to __dataclass_transform__
Model(x=1.0)

# an int is a float, so this also type checks
Model(x=1)

# this does not type check, though
Model(x=Decimal(1))   # "Decimal" is incompatible with "float"

# and this _definitely_ doesn't
Model(x="1")  # "Literal['1']" is incompatible with "float"
```

pydantic [intentionally](https://pydantic-docs.helpmanual.io/usage/models/#data-conversion) coerces all of these into floats at runtime (the float validator is a try/except around `float(x)`). Type checkers flag the last two cases; correctly, since those types _are_ incompatible. 

All of this is known [and documented](https://pydantic-docs.helpmanual.io/visual_studio_code/#strict-errors). It understandably still trips people up (see #3753 in this repo, https://github.com/microsoft/pyright/issues/3678, and certainly others).

(Side note: this is one of my favorite parts of pydantic. I love type safety, but I also love not having to worry whether a poorly-documented external service is returning a string or an int.)
</details>

---

I _think_ that there is a way to type check this in a clearer way than what's currently done. Since the coercion is an intentional decision, I think it's fair to (try to) encode it in the type system. That's what this post does.

This all uses the [descriptor protocol](https://realpython.com/python-descriptors/) to create a new CoercibleFloat type. This behaves like a float at runtime, except it lets you assign non-floats to it (that pydantic will parse into floats) in a static-typing-compatible way. 

Here's the example.

```py
from __future__ import annotations
from decimal import Decimal
from pydantic import BaseModel
from pydantic.validators import float_validator


class CoercibleFloat:
    def __get__(self, instance, owner) -> float: ...

    def __set__(self, instance, value: float | int | str | Decimal | bool): ...

    @classmethod
    def __get_validators__(cls, *args, **kwargs):
        yield float_validator

class AnotherModel(BaseModel):
    y: CoercibleFloat


# all of these type check and work at runtime
AnotherModel(y=4.321)
AnotherModel(y="50")  # previously failed to type check
AnotherModel(y=5)
AnotherModel(y=Decimal(100))  # previously failed to type check

# AnotherModel(y="")  # type checks, but previously failed to type check; always fails at runtime
# AnotherModel(y=None)  # this is a sanity check that doesn't type checking and fails at runtime

# setting the different values still works fine; all of these type check and work at runtime
model = AnotherModel(y=2)
model.y = 55
model.y = "44"
```

What this gets us is the signature of AnotherModel appears _roughly_ like what pydantic accepts:

![image](https://user-images.githubusercontent.com/11623486/180104313-78319e2a-19cf-4805-bd57-ed5226ac5b3f.png)

The primary cost of this is shown with an assignment like `model.y = "abc"`; this has always failed at runtime (if you set `validate_assignment` in the model config, that is), but it previously also failed to type check, and now passes. On the other hand, that's exactly the kind of error that we open ourselves up to when we use pydantic in the non-strict / lax mode—attempting to coerce strings into floats, and sometimes working and sometimes failing. 

---

All of this is supported by the `dataclass_transform` mechanism from [PEP 681](https://peps.python.org/pep-0681/). I think that in practice, this particular feature flew under the radar because it was [discovered](https://mail.python.org/archives/list/typing-sig@python.org/thread/BW6CB6URC4BCN54QSG2STINU2M7V4TQQ/#FC3RIWDWQZTMXPDJT7UM6TKT5BBSXF3V) in a deeply in-the-weeds thread on typing-sig, when Carl Meyer discovered a dataclass/descriptor interaction that appeared to go unnoticed until then.

My hunch is that this kind of typing is more representative of what users want to see from a type checker. If they're using pydantic, it seems likely that they want this coercion. And no matter what, the Python type system can't represent what pydantic is doing, so this can be alternately viewed as another attempt to approximate the pydantic behavior.

There are also some edge cases that I haven't properly figured out, but I think this is a reasonable proof of principle to share and get feedback on.

<details>
<summary>Interaction with default values</summary>

```py
# imports, CoercibleFloat same as above

class Model(BaseModel):
    y: CoercibleFloat = 5  # Literal[5] is incompatible with CoercibleFloat
    y2: float = 5  # however, this works
    z: CoercibleFloat = CoercibleFloat(5)  # Expected no arguments to "CoercibleFloat" constructor
    w: CoercibleFloat = Field(default=5)  # this works, though
```
</details>

## So what am I asking?
First, and most obviously, **is anything egregiously wrong with this post?** Have I overlooked something? Is my understanding of any of this fundamentally flawed?

But secondly, I'd like to propose **encoding these in types that can be used as model annotations,** since the distinction between strict & lax mode will be made more intentional. Concretely, the [conversion table in the v2 blog post](https://pydantic-docs.helpmanual.io/blog/pydantic-v2/#conversion-table) specifies that in "lax" mode, pydantic will (attempt to) coerce any of (float, int, str, Decimal, bool) into a float.

As mentioned earlier, this is imperfect. There's no way to express "strings that look like floats" in the Python type system. What this means in practice is that we're trading spurious errors on things like `Model(float_field="0.01")` for missed errors on things like `Model(float_field="definitely not a float")`. But this tradeoff is roughly what is happening in practice if you're using "lax" mode, anyway.

There's nothing stopping users from doing this right now (I will continue to mess around with this, too), but I think that people in this community generally appreciate the value of type checking and want tools and libraries to work nicely with type checkers. I'm a big fan of tests & types as documentation, which are less likely to desync with the code than human-written docs are.

**Comment by samuelcolvin** [comment]
First of all, thank you so much for the detailed write-up. I think this is really exciting.

A few thoughts:

* Yes, I think what you've summarised here is correct
* I think we should definitely consider this for V2 (although it would be applicable to V1 as well, I don't want to start on this in V1, just so we can move ahead with V2 asap)
* we could perhaps combine descriptors with some constraints - e.g. `CoercibleFloat[0, 100]` would indicate `0 < v < 100`, `CoercibleFloat[0, ..]` would mean `PositiveFloat`
* Maybe we should call it `LaxFloat` not "CoercibleFloat"?
* even if it doesn't work in all scenarios, helping with vscode/pyrigth is a good start if we can avoid adverse effects with other checkers

There are a number of problems with descriptors that I've identified from a brief experiment (I'd love it if anyone can correct me on any of these points!):
* they're not supported by mypy - ignoring `dataclass_transform` issue, I tried your example with mypy and a vanilla dataclass, unfortunately I get `Argument "x" to "Foobar" has incompatible type "float"; expected "CoercibleFloat"` (to be clear - this case does work with pyright)
* Similarly they don't work with pycharm's type hints, or pyre
* even with pyright, they're not supported as function arguments OR as attributes for `TypedDict`s

The problem is that if we implement a `LaxFloat` type which indicates a field should be a lax float and optionally includes some bounds (a bit like `confloat` now), we have two options of how to define it:

### Option 1

```py
if TYPE_CHECKING:
    LaxFloat = float
else:
    class LaxFloat:
        ....
```

* input is `float`: all :heavy_check_mark:
* input is `decimal` or valid `str`: all :skull: (what's the code for a cross in ghfmd?)
* output is `float`: all :heavy_check_mark:

### Option 1

You're about "descriptors" approach.

* input is `float`: pyright / VS code :heavy_check_mark:, mypy pycharm pyre :skull:
* input is `decimal` or valid `str`: pyright / VS code :heavy_check_mark:, mypy pycharm pyre :skull:
* output is `float`: pyright / VS code :heavy_check_mark:, mypy pycharm pyre :skull:

Since these "descriptors" are not supported by mypy, pycharm or pyre I fear we might end up having to go wit the first approach.

---

I guess the best long term approach would be to engage with the typing community, gain broad support for a way to define lax or coercible types which would work in dataclasses, `dataclass_transform`s, function arguments and `TypedDict`s. Then write (and re-write) a PEP to get it fully supported. This would 1) be slow, 2) require a significant effort and time commitment which I'm not prepared to commit to at this time.

**Reply by tuchandra** [reply]
Thanks so much for the reply. I appreciate your openness to the idea and detailed response, even though it seems like there are too many gaps to overcome with something like this.

Descriptors not being supported by function arguments is interesting—especially since that's "just" Python, not a typing question or issue. I don't think I realized that descriptors were _only_ supported as class members, but given that they're defined by having get/set/del, that makes sense on reflection.

I definitely didn't realize that pyright was the only one of those four type checkers to support this kind of pattern, though. It's unsurprising given how complicated this kind of thing is, but a little discouraging, all things considered.

Yeah; I certainly agree with your long-term approach. `TypedDict`s are a pain point of mine, too (I'd love anonymous ones, and a way to specify the key/value typing without the rest of the dict behavior, for things like pandas `DataFrame`s). I've followed the typing PEPs closely and, yeah, it's a massive amount of effort to get them over the finish line (again, understandably, but still hard).
